core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java

```diff
@@ -102,9 +102,17 @@ public Schema record(Schema record, List<String> names, Iterable<Schema.Field> s
 
       } else {
         Preconditions.checkArgument(
-            field.isOptional() || MetadataColumns.metadataFieldIds().contains(field.fieldId()),
-            "Missing required field: %s",
+            (field.isRequired() && field.initialDefault() != null)
+                || field.isOptional()
+                || MetadataColumns.metadataFieldIds().contains(field.fieldId()),
+            "Missing required field that doesn't have a default value: %s",
             field.name());
+        // If the field from Iceberg schema has initial default value, we just pass and don't
+        // project it to the avro file read schema with the generated _r field name,
+        // the default value will be directly read from the Iceberg layer
+        if (field.initialDefault() != null) {
```

Comment:
Why does the case for a `null` default add a field while this one does not? That seems strange to me. I think this should either add a fake field in both cases or not.

I think the question hinges on why the fake field is added for the other case. I think it has something to do with the record schema needing to match the Iceberg schema exactly -- the record schema should not be different. If that's the case then it should be added for the non-null initial default.

Reply:
I see, I originally thought using `continue` can just be an implementation trick, since although the output avro schema of the buildavroprojection process is lacking the field, it will be taken care of when they are being compared again in the `ValueReader`.

I think I can make the code logic more consistent by still creating a fake field with some other suffix like "_d" instead of "_r". The "_d" field will not be projected in the real file either, and the default value read is still taken care of by `ValueReader`. WDYT?

Reply:
What is the purpose of changing the name of the placeholder field that is added? Just use the same code that's already here.